packages/core/src/validate/query.ts

```diff
@@ -144,13 +146,32 @@ export function isUpgradeSafe(data: ValidationData, version: Version): boolean {
   return getErrors(dataV3, version).length == 0;
 }
 
-export function inferProxyKind(data: ValidationData, version: Version): ProxyDeployment['kind'] {
+export async function inferProxyKind(
```

Comment:
This seems reasonable.  It isn't really needed in `deployProxy`, and the main purpose was to give a helpful error when attempting to upgrade a beacon proxy through `upgradeProxy` or `prepareUpgrade`.  But those functions could just call `isBeaconProxy()` directly.

Another reason for this original approach though was so the Truffle plugin could also detect previously deployed beacon proxies when the user is trying to upgrade them (and then give an error about mismatching `kind`), rather than infer the wrong `kind`.  Although of course this PR does not cover Truffle and perhaps inferring the wrong `kind` is not much of a problem if it fails gracefully otherwise.